File: advall-python/d20.py
path = "d20_input.txt"

def parseInput(path):
    ranges = []
    for line in open (path, "r"):
        s = line.rstrip().split("-")
        s = (int(s[0]), int(s[1]))
        ranges.append(s)
    return ranges

def solve(ranges):
    i = 0
    part1Sol = -1
    part2Sol = 0
    ranges.sort()
    while i < 4294967296:
        if len(ranges) > 0:
            fstRange = ranges.pop(0)
        else:
            part2Sol += (4294967296 - i)
            break

        if (i < fstRange[0]):
            if (part1Sol < 0):
                part1Sol = i
            part2Sol += 1
            i += 1
            ranges.insert(0, fstRange)
        elif (i >= fstRange[0]) and (i <= fstRange[1]):
            i = fstRange[1] + 1

    return (part1Sol, part2Sol)


solutions = solve(parseInput(path))
print "Part one: " + str(solutions[0])
print "Part two: " + str(solutions[1])

File: advall-haskell/d20.hs
import Data.List.Split
import Data.List

solve1 :: Integer -> [(Integer, Integer)] -> Integer
solve1 floor ((h,t):xs)
    | floor >= h = solve1 ((max floor t) + 1) xs
    | otherwise = floor

solve2 :: Integer -> Integer -> [(Integer, Integer)] -> Integer
solve2 floor roof [] = roof - floor + 1
solve2 floor roof ((h,t):xs)
    | floor < h = (h - floor) + solve2 h roof ((h,t):xs)
    | (floor >= h) && (floor <= t) = solve2 (t + 1) roof xs
    | otherwise = solve2 floor roof xs

parseInput :: String -> [(Integer, Integer)]
parseInput s = sort (map parseInput' (map (splitOn "-") (lines s)))

parseInput' :: [String] -> (Integer, Integer)
parseInput' [x,y] = (read x, read y)

main = do
    input <- readFile "d20_input.txt"
    let answer1 = solve1 0 (parseInput input)
    putStrLn $ "Part one: " ++ show answer1
    let answer2 = solve2 0 4294967295 (parseInput input)
    putStrLn $ "Part two: " ++ show answer2